Fix frustrum culling for sprites with pivot point

ekharkunov opened the pull request:
Fixed culling for sprites with pivot points. Also implemented some optimization on how geometry data calculated and stored between batches. So no it should be a bit faster in terms of calculation but consume some memory to store cached information.

Fixes #9917 

### Technical changes
`TexturesData` completely removed because all data can be retrieved from `SpriteComponent`.
In some inner function replace 3 or 4 function arguments with 1 SpriteComponent and access data inside the function.

For animation data cache used the same approach as it was implemented in `ResourceChunkCache`. The only difference is cache eviction rule. I counted how every sprite world update. Every time when cache entry is retrieved from the cache - updated special field in cache entry (like timestamp) which stores what engine tick entry was read at.
2 structures to work with cache are used:
* hastable to fast access to cache data by key
* double linked list to implement Least Recently Used cache

When cache entry is accessed - record in LRU cache moved to head. Then when cache eviction rule is checked - I iterate from the end of the LRU until I found the first cache entry which should not be evicted.

P.S. Checked rotated image in atlas with pivot point - calculation is already correct.

Measurement:
Project contains 2 collection. 5 times collection will be changed (one collection unloaded, second - load). Every collection contains a set of game objects. First collection contains 4 types of game object. Every game object has their own texture with one animation. Second collection contains 5 types of game object. Every object contains their own texture with one sprite.
Every collection loads, spawn 6k of random objects, then remove 5k, then spawn 5k and load next collection.

Memory
|Build|Peak (Mb)| Average (Mb)|
|-----|----------|--------------|
|base |110,38     |~92                |
|cache|100,27    |~83,5             |

Time
base:  CompSpriteRender -2,68s, CompSpriteUpdate - 820 ms
cache: CompSpriteRender - 2,92s, CompSpriteUpdate - 901ms

Base:

<img width="1236" alt="Screenshot 2025-05-26 at 00 57 59" src="https://github.com/user-attachments/assets/fd0d03c9-114f-48ed-a6b6-a4ec78fa2cdd" />
<img width="1267" alt="Screenshot 2025-05-26 at 00 57 40" src="https://github.com/user-attachments/assets/b25ed8de-f807-4cde-b832-2624901fa2e4" />

Cache:
<img width="1229" alt="Screenshot 2025-05-26 at 01 02 51" src="https://github.com/user-attachments/assets/50fb253b-f3af-468f-ab63-877539666046" />
<img width="1228" alt="Screenshot 2025-05-26 at 01 03 38" src="https://github.com/user-attachments/assets/811e6c78-89c0-4e52-81dc-4d3c16b65c56" />

Another set of optimizations was done in https://github.com/defold/defold/pull/10736

## PR checklist


* [x] Code
	* [x] Add engine and/or editor unit tests.
	* [x] New and changed code follows the overall code style of existing code
	* [x] Add comments where needed
* [x] Documentation
	* [x] Make sure that API documentation is updated in code comments
	* [x] Make sure that manuals are updated (in github.com/defold/doc)
* [x] Prepare pull request and affected issue for automatic release notes generator
	* [x] Pull request - Write a message that explains what this pull request does. What was the problem? How was it solved? What are the changes to APIs or the new APIs introduced? This message will be used in the generated release notes. Make sure it is well written and understandable for a user of Defold.
	* [x] Pull request - Write a pull request title that in a sentence summarises what the pull request does. Do not include "Issue-1234 ..." in the title. This text will be used in the generated release notes.
	* [x] Pull request - Link the pull request to the issue(s) it is closing. Use on of the [approved closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).
	* [x] Affected issue - Assign the issue to a project. Do not assign the pull request to a project if there is an issue which the pull request closes.
	* [x] Affected issue - Assign the "breaking change" label to the issue if introducing a breaking change.
	* [x] Affected issue - Assign the "skip release notes" is the issue should not be included in the generated release notes.

----------

Example of a well written PR description:

1. Start with the user facing changes. This will end up in the release notes.
1. Add one of the GitHub approved closing keywords
1. Optionally also add the technical changes made. This is information that might help the reviewer. It will not show up in the release notes. Technical changes are identified by a line starting with one of these:
   1. `### Technical changes` 
   1. `Technical changes:`
   2. `Technical notes:`

```
There was a anomaly in the carbon chroniton propeller, introduced in version 8.10.2. This fix will make sure to reset the phaser collector on application startup.

Fixes #1234

### Technical changes
* Pay special attention to line 23 of phaser_collector.clj as it contains some interesting optimizations
* The propeller code was not taking into account a negative phase.
```

ekharkunov commented on engine/gamesys/src/gamesys/components/comp_sprite.cpp:
```diff
@@ -65,8 +65,27 @@ namespace dmGameSystem
+
     struct SpriteComponent
     {
+        TexturesData                m_TexturesData; // store precalculated gemetry data that used in different places during the frame
```
Stores textures data to avoid double calculation in `UpdateTextureDataAndCounts` (ex `UpdateVertexAndIndexCount`) and in `CreateVertexData`. So calculate once in `UpdateTextureDataAndCounts` and use that result in `CreateVertexData`.

JCash replied:
I was a bit surprised to see this in a PR regarding the pivot point?

Also, I'd expect some numbers on the trade offs for this.
E.g. now the sprite component structure becomes much larger.
I.e. what was the size of 32k sprites before/after?
What is the _measured_ performance benefit.
I feel we need those things before making a judgement call, especially as we are currently looking to minimize runtime footprint where available.

ekharkunov replied:
Adding those data to `SpriteComponent` is more about to solve current issue.
Pivots we calculate only inside `CreateVertexData`. Call chain: `RenderListDispatch`->`RenderBatch`->`CreateVertexData`. So `RenderListFrustumCulling` is calling before pivot points will be calculated.
So I guess more correct solution is to precalculate pivot points and stored it in `SpriteWorld` like as done for `m_BoundingVolumes`.

ekharkunov commented on engine/gamesys/src/gamesys/components/comp_sprite.cpp:
```diff
@@ -1218,14 +1213,12 @@ namespace dmGameSystem
             float sp_height = component->m_Size.getY();
 
-            // Get the correct animation frames, and other meta data
-            ResolveAnimationData(&textures, component->m_CurrentAnimation, component->m_CurrentAnimationFrame);
```
No need to do calculation here any more.

ekharkunov commented on engine/gamesys/src/gamesys/components/comp_sprite.cpp:
```diff
@@ -1737,11 +1736,12 @@ namespace dmGameSystem
     }
 
-    static void UpdateVertexAndIndexCount(SpriteWorld* sprite_world, dmRender::HRenderContext render_context)
+    static void UpdateTextureDataAndCounts(SpriteWorld* sprite_world, dmRender::HRenderContext render_context)
```
Update function name to be more meaningful to new functionality.

ekharkunov commented on engine/gamesys/src/gamesys/components/comp_sprite.cpp:
```diff
@@ -1962,7 +1962,25 @@ namespace dmGameSystem
+                        pivot_y = geometry->m_PivotY;
+                    }
+
+                    trans = world_transform * Point3(-pivot_x, -pivot_y, 0.0f);
```
Main fix. I looked that `m_WorldsPosition` in `RenderListEntry` used only for sorting and for visibility check so it's ok to set `m_WorldPosition` to value with pivot points.
Let me know if it's not ok.

JCash replied:
Structurally, it feels a bit off, since the other application of the pivot is for vertex calculations, but here a special case is added on the object's world position.

So now we apply pivot in "two" places.

Also, the pivot isn't about changing the position of the component, but shifting the visual representation.
So this change should go into the corresponding place for the vertex calculation.

ekharkunov commented on engine/gamesys/src/gamesys/components/comp_sprite.cpp:
```diff
@@ -65,8 +65,27 @@ namespace dmGameSystem
+
     struct SpriteComponent
     {
+        TexturesData                m_TexturesData; // store precalculated gemetry data that used in different places during the frame
```
Probably we can cache only geometries and left double calculation 'as-is' to avoid grows of memory footprint. Let's discuss.

JCash reviewed (changes requested):
Please split change into two PR's.
Also use the pivot when generating the vertices, to leave the object position as-is.

ekharkunov committed:
WIP pivots

ekharkunov committed:
Merge remote-tracking branch 'origin/dev' into issue-9917-sprite-pivot-culling

ekharkunov committed:
Moved double linked list to dlib.
Implemented cache eviction for animation data in sprite component

ekharkunov committed:
Fixed cache eviction condition in case of uint overflow.
Fixed sprite culling with pivot points

ekharkunov commented on engine/gamesys/src/gamesys/components/comp_sprite.cpp:
```diff
@@ -65,6 +66,31 @@ namespace dmGameSystem
+    const uint32_t MAX_TEXTURE_COUNT = dmRender::RenderObject::MAX_TEXTURE_COUNT;
+    const uint8_t CACHE_EVICTION_FRAMES = 10; // how many frames cache entry is stored before eviction
+
+    struct AnimationData
```
Use the same approach as in ResourceChunkCache.

JCash replied:
I'd also like a code comment about how the animation data usage is intended to work and why.
This component is getting quite complex, and I think we'll need more documentaion moving on.

ekharkunov commented on engine/gamesys/src/gamesys/components/comp_sprite.cpp:
```diff
@@ -65,6 +66,31 @@ namespace dmGameSystem
+        dmhash_t                        m_AnimationID;                      // The animation of the driving atlas
+        uint32_t                        m_Frames[MAX_TEXTURE_COUNT];        // The resolved frame indices
+        uint32_t                        m_LastAccessTick;
+        uint32_t                        m_CacheKey;
```
Added cache key as a field because in case when cache entry will be removed because of cache eviction we should know the key to correctly update hashtable.

ekharkunov commented on engine/gamesys/src/gamesys/components/comp_sprite.cpp:
```diff
@@ -65,6 +66,31 @@ namespace dmGameSystem
+        // Used after resolving info from all textures
+        dmhash_t                        m_AnimationID;                      // The animation of the driving atlas
+        uint32_t                        m_Frames[MAX_TEXTURE_COUNT];        // The resolved frame indices
+        uint32_t                        m_LastAccessTick;
```
Use engine tick as a marker for cache eviction.
I can use for ticks counting `uint16_t` instead of `uint32_t` because overflow case correctly handles.

ekharkunov commented on engine/gamesys/src/gamesys/components/comp_sprite.cpp:
```diff
@@ -65,6 +66,31 @@ namespace dmGameSystem
+    {
+        dmHashTable32<AnimationData*> m_Cache;
+        dmDoubleLinkedList::List      m_LRU;
+        uint32_t                      m_CurrentEngineTick;
```
That filed updates every call of `CompSpriteUpdate`

ekharkunov commented on engine/gamesys/src/gamesys/components/comp_sprite.cpp:
```diff
@@ -137,22 +162,6 @@ namespace dmGameSystem
         uint8_t                             m_ReallocBuffers : 1;
     };
 
-    struct TexturesData
```
Completely removed that structure because all information contains inside `SpriteComponent`

ekharkunov commented on engine/resource/src/resource_chunk_cache.cpp:
```diff
@@ -16,28 +16,17 @@
 #include <stddef.h>  // offsetof
 #include <algorithm> // lower_bound
 
-struct DLListNode
```
Moved to dlib to use in comp_sprite.

JCash commented on engine/gamesys/src/gamesys/components/comp_sprite.cpp:
```diff
@@ -1844,6 +1942,8 @@ namespace dmGameSystem
          */
 
         SpriteWorld* world = (SpriteWorld*)params.m_World;
+        UpdateCache(world);
```
Perhaps rename to UpdateAnimationDataCache() ?

ekharkunov replied:
Renamed

JCash commented on engine/dlib/src/dlib/double_linked_list.h:
```diff
@@ -0,0 +1,42 @@
+// Copyright 2020-2025 The Defold Foundation
```
The resource chunk cache has some unit tests, but now that the api is formalized, we need to add tests for this container too.

ekharkunov replied:
Added test_double_linked_list.cpp

JCash commented on engine/gamesys/src/gamesys/components/comp_sprite.cpp:
```diff
@@ -1182,7 +1230,46 @@ namespace dmGameSystem
+    static AnimationData* GetOrCreateAnimationData(SpriteWorld* sprite_world, const SpriteComponent* component)
+    {
+        // component, component->m_MixedHash, component->m_CurrentAnimation, component->m_CurrentAnimationFrame
+        HashState32 state;
+        dmHashInit32(&state, false);
+        dmHashUpdateBuffer32(&state, &component->m_MixedHash, sizeof(component->m_MixedHash));
+        dmHashUpdateBuffer32(&state, &component->m_CurrentAnimation, sizeof(component->m_CurrentAnimation));
+        dmHashUpdateBuffer32(&state, &component->m_CurrentAnimationFrame, sizeof(component->m_CurrentAnimationFrame));
+        uint32_t hash = dmHashFinal32(&state);
```
As the animation doesn't generally change each update, it might be appropriate to update this when the m_CurrentAnimation/m_CurrentAnimationFrame changes, and then use a m_AnimationHash to do this lookup.

JCash commented on engine/gamesys/src/gamesys/components/comp_sprite.cpp:
```diff
@@ -65,6 +66,31 @@ namespace dmGameSystem
+        bool                                        m_UsesGeometries;
+    };
+
+    struct AnimationDataCache
```
It would be good with a note on _why_ we want to use a cache at all, as opposed to using a ref count for the structure.

JCash reviewed (changes requested):
Overall good. Only smaller things needed
* Can pre calculate the animation state hash
* Needs some code comments
* Need a unit test for the new container
* What's the performance comparison before/after this change?

ekharkunov committed:
Added animation data hash update on demand.
Added tests for double linked list
Updated documentation

ekharkunov committed:
Fixed crash when tries to remove list node which haven't be added before

ekharkunov committed:
Added code comments

ekharkunov committed:
Initialize ListNode fields with zero

ekharkunov committed:
Move cache entry only once per frame

ekharkunov committed:
Merge remote-tracking branch 'origin/dev' into issue-9917-sprite-pivot-culling

ekharkunov committed:
Fix compilation after merge

ekharkunov committed:
Various sprite component optimizations (#10736)

* Various sprite component optimizations (not working)

Use smaller types for dynamic attribute index and animation id
Cache vertex and indices count in AnimationData.

Cache CanUseQuads flag

Fixed case when CompSpriteUpdate can be called several times as result vertex buffer stay unallocated

Fixed GetAnimationFrameCount when animation is not playing

Fixed GetCursor tests

Repack sprite component

ekharkunov committed:
Use dmHashReverseSafe64Alloc for reverse hash
